server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java

```diff
@@ -197,20 +200,41 @@ private PublishWithJoinResponse handleIncomingPublishRequest(BytesTransportReque
                         incomingState.stateUUID(),
                         request.bytes().length()
                     );
-                    final PublishWithJoinResponse response = acceptState(incomingState);
-                    lastSeenClusterState.compareAndSet(lastSeen, incomingState);
-                    return response;
+                    acceptState(
+                        incomingState,
+                        transportChannel,
+                        (acceptedState) -> lastSeenClusterState.compareAndSet(lastSeen, acceptedState)
+                    );
                 }
             }
         } finally {
             IOUtils.close(in);
         }
     }
 
-    private PublishWithJoinResponse acceptState(ClusterState incomingState) {
+    /**
+     * Delegate to cluster-coordination thread to apply received state
+     *
+     * @param incomingState The received cluster state
+     * @param transportChannel The channel on which it was received
+     * @param onSuccess The action to perform if the new state is accepted
+     */
+    private void acceptState(ClusterState incomingState, TransportChannel transportChannel, Consumer<ClusterState> onSuccess) {
         assert incomingState.nodes().isLocalNodeElectedMaster() == false
             : "should handle local publications locally, but got " + incomingState;
-        return handlePublishRequest.apply(new PublishRequest(incomingState));
+        clusterCoordinationExecutor.execute(ActionRunnable.supply(new ActionListener<>() {
+
+            @Override
+            public void onResponse(PublishWithJoinResponse publishWithJoinResponse) {
+                onSuccess.accept(incomingState);
+                transportChannel.sendResponse(publishWithJoinResponse);
+            }
```

Comment:
I considered using `ActionListener.runBefore(..., () -> onSuccess.apply(initialState))` but I think just putting the call to `onSuccess` inline in `onResponse` is a bit clearer (it's easy to miss the `runBefore` wrapping IMO, but I'm still new to these utilities)
I know using `runBefore` would behave better in the event onSuccess threw an exception, but it's only setting an `AtomicReference` so I don't think that's even possible the way things are.
Open to feedback on that.

Reply:
If we passed in an `ActionListener` rather than a `TransportChannel` then the `runBefore` would not be so lost in the noise I think.